lib/Glimmer.pm

```diff
@@ -0,0 +1,494 @@
+	}
+    }
+    close(TRAIN) || die "Could not close $tmp_train";
+    close(TBL)   || die "Could not close $training_tbl";
+    
+    if (($_ = `grep -c "^>" $tmp_train`) && ($_ =~ m/^\s*(\d+)/)) {
+	print STDERR "\nExtracted $1 training sequences\n\n" if $verbose;
+    } 
+    else {
+	die "\nCould not extract any training sequences";
+    }
+    
+    #+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    #... Build ICM (interpolated context model)
+    #-----------------------------------------------------------------------
+    print STDERR ("Building interpolated context model ---\n",
+		  "   output in $tmp_model\n\n"
+		 ) if $verbose;
+    
+    if (-s "$tmp_model") {
+	system("rm -f $tmp_model") && die "Could not remove $tmp_model";
+    }
+    
+    run($verbose, "build-icm -r $tmp_model < $tmp_train");
+    
+    
+    
+    #+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    #... First GLIMMER pass
+    #-----------------------------------------------------------------------
+    print STDERR "Running first GLIMMER pass\n" if $verbose;
+    run($verbose, "glimmer3 $glimmeropts $genetic_code_switch $contigs_file $tmp_model $tmp_prefix.pass_1");
+    
+    
+    
+    #+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    #... Extract upstream regions and START codon counts
+    #-----------------------------------------------------------------------
+    my $tmp_predictions_pass_1 = "$tmp_prefix.pass_1.predict";
+    print STDERR "\nExtracting upstream regions and START counts from $tmp_predictions_pass_1\n"
+	if $verbose;
+    
+    open(PREDICT,  "<$tmp_predictions_pass_1")
+	|| die "Could not read-open $tmp_predictions_pass_1";
+    
+    my $tmp_upstream = "$tmp_prefix.upstream";
+    open(UPSTREAM, ">$tmp_upstream")
+	|| die "Could not write-open $tmp_upstream";
+    
+    my %start_counts;		#...Hash to hold START codon counts
+    $start_counts{0}   = 0;
+    $start_counts{atg} = 0;
+    $start_counts{gtg} = 0;
+    $start_counts{ttg} = 0;
+    
+    $contig_id = qq();
+    while(defined($entry = <PREDICT>)) {
+	chomp $entry;
+	
+	if ($entry =~ m/^>(\S+)/) {
+	    $contig_id = $1;
+	    next;
+	}
+	elsif ($entry =~ m/^(\S+)\s+(\d+)\s+(\d+)/) {
+	    my ($id, $beg, $end) = ($1, $2, $3);
+	    
+	    my $sign = ($beg < $end) ? +1 : -1;
+	    
+	    my $end_start   = $beg + $sign * 2;
+	    my $start_codon = &get_dna_seq($contig_id, $beg, $end_start, $len_of, $seq_of);
+	    
+	    ++$start_counts{0};
+	    ++$start_counts{$start_codon};
+	    
+	    my $up_id = $id;
+	    $up_id =~ s/^orf/ups/o;
+	    my $up_beg = $beg - $sign * 25;
+	    my $up_end = $beg - $sign;
+	    
+	    my $up_seq;
+	    if ($up_seq = &get_dna_seq($contig_id, $up_beg, $up_end, $len_of, $seq_of)) {
+		&display_id_and_seq($id, \$up_seq, \*UPSTREAM);
+	    }
+	}
+	else {
+	    die "Could not parse prediction $.: $entry";
+	}
+    }
```

Comment:
Please consider breaking down this function into smaller functions -- e.g. this could be a single function. It makes it much easier to test, and much easier to read and understand. When you're writing code, you should always be thinking, "How can I test this?". A long function with lots of potential failure points and lots of input/output is going to be much more difficult to test (and thus to diagnose bugs).